rfcs/0000-c-unwind-abi.md

```diff
@@ -0,0 +1,273 @@
+[summary]: #summary
+
+Introduces a new ABI string, `"C unwind"`, to enable unwinding from other
+languages (such as C++) into Rust frames and from Rust into other languages.
+
+Additionally, defines the behavior for a limited number of previously-undefined
+cases when an unwind operation reaches a Rust function boundary with a
+non-`"Rust"`, non-`"C"` ABI.
+
+This RFC does not define the behavior of `catch_unwind` in a Rust frame being
+unwound by a foreign exception.
+
+# Motivation
+[motivation]: #motivation
+
+There are some Rust projects that need cross-language unwinding to provide
+their desired functionality. One major example is WASM interpreters, including
+the Lucet and Wasmer projects.
+
+There are also existing Rust crates (notably, wrappers around the `libpng` and
+`libjpeg` C libraries) that rely on compatibility with the native exception
+handling mechanism in GCC, LLVM, and MSVC.
+
+Additionally, there are libraries such as `rlua` that rely on `longjmp` across
+Rust frames; on Windows, `longjmp` is implemented via unwinding.
+
+The desire for this feature has been previously discussed on other RFCs,
+including #2699 and #2753.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring an external function that may unwind, such as an entrypoint to a
+C++ library, use `"C unwind"`:
+
+```
+extern "C unwind" {
+  fn may_throw();
+}
+```
+
+Rust functions that call a possibly-unwinding external function should either
+use the default Rust ABI (which can be made explicit with `extern "Rust"`) or
+the `"C unwind"` ABI:
+
+```
+extern "C unwind" fn can_unwind() {
+  // ...
+```
+
+Using the `"C unwind"` ABI to "sandwich" Rust frames between frames from
+another language (such as C++) allows an exception initiated in a callee frame
+in the other language to traverse the intermediate Rust frames before being
+caught in the caller frames. I.e., a C++ exception may be thrown,
+cross into Rust via an `extern "C unwind"` function declaration, safely unwind
+the Rust frames, and cross back into C++ (where it may be caught) via a Rust
+`"C unwind"` function definition.
+
+Conversely, languages that support the native unwinding mechanism, such as C++,
+may be "sandwiched" between Rust frames, so that Rust `panic`s may safely
+unwind the C++ frames, if the Rust code declares both the C++ entrypoint and
+the Rust entrypoint using `"C unwind"`.
+
+## Forced unwinding
+[forced-unwinding]: #forced-unwinding
+
+This is a special kind of unwinding used to implement `longjmp` on Windows and
+`pthread_exit` in `glibc`. A brief explanation is provided in [this Inside Rust
+blog post][inside-rust-forced]. This RFC distinguishes forced unwinding from
+other types of foreign unwinding.
+
+Since language features implemented using forced unwinding on some platforms
+use other mechanisms on other platforms, Rust code should not rely on forced
+unwinding to invoke destructors (calling `drop` on `Drop` types). External
```

Comment:
This seems to imply that forced unwinding accross frames with `Drop` types is allowed, rather than being UB as said below.

Reply:
When destructors don't run, that's undefined behavior. But in the case where `"C unwind"` is used, forced unwinding across destructors will in fact invoke destructors under this proposal, so it's _technically_ not UB according to this proposal.

@Amanieu @nikomatsakis @joshtriplett This seems like an oddity to me; the existence, or not, of UB should generally be determined by the language virtual machine model, not by the platform. So even though we expect this case to work "in practice", it seems that the language spec should leave it UB; do you agree? (This would also resolve the concern that the proposal has a case where changing from `panic=unwind` to `panic=abort` introduces UB.)

Reply:
There is some trickiness here based on whether we define the behavior in terms of *forced exceptions* or in terms of *particular, well-known C functions*.

In practice, it seems like forced exceptions seem to be "highly correlated" with abrupt jumps, like longjmp and pthread-exit, that are also implemented in different ways across platforms. Let's them "abrupt-jump" operations.

I think the core invariant is:

* Let's define a Rust stack frame as "inert" if it it has no pending destructors and does not correspond to the `catch_panic` intrinsic. 
* It is UB to unwind or otherwise deallocate a Rust stack frame without running its destructors.
* Forced unwind across a "C" ABI boundary is:
    * not a problem if all Rust stack frames are inert
    * UB otherwise (because it would cause Rust destructors not to execute)
* Forced unwind across a "C unwind" ABI boundary is:
    * OK if all stack frames are "inert"
    * "Not Yet Specified" otherwise, because we are still working on what it means for a foreign exception to propagate across non-inert Rust frames

We could then say that the implementation of operations like longjmp and pthread-exit varies by platform. The only **portable way** to use those functions that is not UB is 

* with the "C" or "C unwind" ABIs and to only drop inert Rust frames

In particular, on non-Windows platforms, longjmp will immediately deallocate frames, which is UB unless those frames are inert. On Windows, longjmp will trigger a forced exception, and the interaction of foreign exceptions with non-inert frames is Not Yet Specified.

Note 1: recall that "Not Yet Specified" just means UB that we maybe intend to change to something defined later (but it's UB until we do so).

Note 2: I believe that C++ just kind of jumps to the final point, and directly defines longjmp across C++ frames as UB if those frames are "pending" (per my definition above). I think it's ok to define the core operations and say that the behavior of longjmp etc ultimately depends on what core operations are used to implement it.

Thoughts?

Reply:
Hm, I thought we had agreed that there's no reason to leave the foreign-exceptions case not-yet-specified. The specification (in this RFC) seems straightforward to me; if there is a remaining doubt about the specifics of the specification, then I'd be okay with leaving it not-yet-specified, but I don't know that that's the case.

Do you want to include new definitions for destructor-free, catch-free frames and for frames that don't fit those criteria in the RFC itself? If so, I'm not quite sold on the "inert"/"pending" terminology. Those words aren't the inverse of each other in common language, and "inert" doesn't strike me as an obvious analogy. I think perhaps "disposable" might be better, since the frames can be "disposed" without side-effects. Or, we can simply apply Rust's trait name to the frames themselves: frames with `Drop` locals are "drop" frames. (The nonstandard verb-as-adjective syntax is intentional, to signal that it's a technical term.)

A Rust compiler (maybe not `rustc` specifically) could even emit code to avoid invoking destructors in the case of forced-unwind; that would make thread-termination and `longjmp` have "real" (unpredictable/dangerous in practice) UB across such "pending"/"non-disposable" frames on all platforms. Should the RFC specifications (or the reference, or a future language spec) prohibit this? I'm not sure there would be a benefit to doing so.

This leaves me with the opinion that we should say "forced unwinding across drop frames is always UB." That way, we can make the equivalent statement that "pthread-cancelling and `longjmp` are UB when they cross drop frames, on all platforms."

Reply:
I've revised the RFC to this effect. I think it simplifies the specification a fair amount, actually.

Reply:
@nikomatsakis marking this resolved for now; please un-resolve if you think this is the wrong decision. (@Amanieu you may also disagree with my decision.)